### ISSUE TITLE
Numerical stability issues in gradient for log1mexp

### ISSUE THREAD

**Issue by rmanhaeve** [ML-KULeuven/klay#17]
There's some numerical issues in numerical stability in log1mexp. Here's a test to verify this:
```
def test_log1mexp_gradient_stability():
	test_cases = [
		torch.tensor([-0.01, -0.1, -1.0, -10.0, -100.0], dtype=torch.float32),
		torch.tensor([-1000.0], dtype=torch.float32),
		torch.tensor([-1e-10], dtype=torch.float32),  # Very close to 0
	]

	for i, x in enumerate(test_cases):
		x_test = x.clone().requires_grad_(True)


		output = log1mexp(x_test)

		assert torch.isfinite(output).all(), f"Output should be finite for case {i + 1}"

		loss = output.sum()
		loss.backward()

		assert torch.isfinite(x_test.grad).all(), f"Gradient should be finite for case {i + 1}"
```
A fix is using clamp with eps instead of summing:

```
def log1mexp(x):
    """
    Numerically accurate evaluation of log(1 - exp(x)) for x < 0.
    See [Maechler2012accurate]_ for details.
    https://github.com/pytorch/pytorch/issues/39242
    """
    mask = CUTOFF < x  # x < 0
    return torch.where(
        mask,
        (torch.clamp(-x.expm1(), min=EPS)).log(),
        (torch.clamp(-x.exp(), min=EPS)).log1p(),
    )
```
